go/cmd/server/logging/logging.go

```diff
@@ -0,0 +1,391 @@
+	fvs := make(map[string]*model.FeatureView)
+	odFvs := make(map[string]*model.OnDemandFeatureView)
+
+	joinKeys := make([]string, 0)
+	// All joinkeys in the featureService are put in here
+	joinKeysSet := make(map[string]interface{})
+	entityJoinKeyToType := make(map[string]types.ValueType_Enum)
+
+	for _, featureView := range featureViews {
+		fvs[featureView.Base.Name] = featureView
+	}
+
+	for _, onDemandFeatureView := range onDemandFeatureViews {
+		odFvs[onDemandFeatureView.Base.Name] = onDemandFeatureView
+	}
+
+	allFeatureTypes := make(map[string]types.ValueType_Enum)
+	features := make([]string, 0)
+	for _, featureProjection := range featureService.Projections {
+		// Create copies of FeatureView that may contains the same *FeatureView but
+		// each differentiated by a *FeatureViewProjection
+		featureViewName := featureProjection.Name
+		if fv, ok := fvs[featureViewName]; ok {
+			for _, f := range featureProjection.Features {
+				features = append(features, GetFullFeatureName(featureViewName, f.Name))
+				allFeatureTypes[GetFullFeatureName(featureViewName, f.Name)] = f.Dtype
+			}
+			for entityName := range fv.Entities {
+				entity := entityMap[entityName]
+				if joinKeyAlias, ok := featureProjection.JoinKeyMap[entity.JoinKey]; ok {
+					joinKeysSet[joinKeyAlias] = nil
+				} else {
+					joinKeysSet[entity.JoinKey] = nil
+				}
+			}
+		} else if odfv, ok := odFvs[featureViewName]; ok {
+			for _, f := range odfv.Base.Features {
+				// TODO(kevjumba) check in test here.
+				features = append(features, GetFullFeatureName(featureViewName, f.Name))
+				allFeatureTypes[GetFullFeatureName(featureViewName, f.Name)] = f.Dtype
+			}
+		} else {
+			return nil, fmt.Errorf("no such feature view found in feature service %s", featureViewName)
+		}
+	}
+
+	// Only get entities in the current feature service.
+	for _, entity := range entities {
+		if _, ok := joinKeysSet[entity.Name]; ok {
+			joinKeys = append(joinKeys, entity.JoinKey)
+			entityJoinKeyToType[entity.JoinKey] = entity.ValueType
+		}
+	}
+
+	schema := &Schema{
+		Entities:      joinKeys,
+		Features:      features,
+		EntityTypes:   entityJoinKeyToType,
+		FeaturesTypes: allFeatureTypes,
+	}
+	return schema, nil
+}
+
+func GetFullFeatureName(featureViewName string, featureName string) string {
+	return fmt.Sprintf("%s__%s", featureViewName, featureName)
+}
+
+func (s *LoggingService) GetFcos() ([]*model.Entity, map[string]*model.Entity, []*model.FeatureView, []*model.OnDemandFeatureView, error) {
+	odfvs, err := s.fs.ListOnDemandFeatureViews()
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+	fvs, err := s.fs.ListFeatureViews()
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+	entities, err := s.fs.ListEntities(true)
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+	entityMap := make(map[string]*model.Entity)
+	for _, entity := range entities {
```

Comment:
let's maybe move all maps creation here?

Reply:
I don't think there are any other map creations that we need? Only this one, the other entityValueMap needs to be created in GetOnlineFeatures.